### PR TITLE
42 finish authentication

### LINKED ISSUE #42: finish authentification
Change the system from doing register/login activity to doing a class with functions reusable and to modularize the code

### PR DESCRIPTION
**Pull request by abouquet27**
Closes #42 

Creates an account by asking the user its information (birthday, firstname, etc), check the format and send it to the databases

### CODE REVIEW

**Review by bluedrack — CHANGES_REQUESTED** [review]
There is somethings that seem off to me as listed. But make sure that you have the 80% code coverage needed. Maybe add some test to the FirebaseCreateAccount class to achieve the code coverage by checking null value or incorrect phone number or name format.

File: app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    * Method to create a user account in Firebase database.
+    * @param year the user's year of birth
+    * @param month the user's month of birth
+    * @param day the user's day of birth
+    * @param firstname the user's first name
+    * @param lastname the user's last name
+    * @param phone the user's phone number
+    */
+    @Inject
+    override fun postData(year: Int, month: Int, day: Int, firstname: String, lastname: String, phone: String){
```
[the remark this answers is not in the document]
Comment (abouquet27) [reply app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt:51]: If you check the CreateAccountActivity.kt it is already done outside of the function but I can do it inside too

[the remark this answers is not in the document]
Comment (abouquet27) [reply app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt:51]: Well, it was my first plan but after some discussion with Willy, we conclude that it was not useful. Maybe it would an interesting discussion to have together

File: app/src/main/java/com/github/studydistractor/sdp/register/RegisterActivity.kt [hunk 2]
```diff
@@ -158,19 +154,14 @@ class RegisterActivity : Hilt_RegisterActivity() {
...
-//        db.getReference("users").child(userid).setValue(hashMapOf(
-//            "score" to 0,
-//            "birthday" to hashMapOf("day" to 0, "mouth" to 0, "year" to 0),
-//            "firstname" to "test",
-//            "lastname" to "test",
-//            "phone" to "0000000000"
-//        ))
-
+    private fun addUserToTheDatabase() {
+        val intent = Intent(this, CreateAccountActivity::class.java)
```
[the remark this answers is not in the document]
Comment (abouquet27) [reply app/src/main/java/com/github/studydistractor/sdp/register/RegisterActivity.kt:163]: goToRegister() is the method of the LoginActivity that leads to the RegisterActivity(). The addUserToTheDatabase is effectively misleading. I'll change it to createAccount()

File: app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (abouquet27) [reply app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt:52]: As long as I check it before, it does not impact the correctness of the code I think.

File: app/src/main/java/com/github/studydistractor/sdp/account/CreateAccountActivity.kt [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+     * phone number, and birthday. The function also includes a button for validating the
+     * entered data and creating a new account.
+     * The function uses Jetpack Compose for building the UI components, and it relies on the
+     * DatePickerDialog to allow the user to select their birthday. The function also uses
+     * the LocalContext.current property to access the current context of the application.
+     * @return A composable UI screen for creating a new account
+     */
+    @Preview
+    @Composable
+    fun CreateAccScreen() {
```
Comment (aizensoosuke) [inline app/src/main/java/com/github/studydistractor/sdp/account/CreateAccountActivity.kt:72]: Please rename to `CreateAccountScreen` for consistency

File: app/src/main/java/com/github/studydistractor/sdp/account/CreateAccountActivity.kt [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+                    datePicker.show()
+                },
+                colors = ButtonDefaults.buttonColors(backgroundColor = Color.White),
+                modifier = Modifier
+                    .fillMaxWidth()
+                    .padding(vertical = 8.dp)
+                    .testTag("birthday"),
+
+            ) {
+                Icon(Icons.Filled.CardGiftcard, contentDescription = null)
```
Comment (aizensoosuke) [inline app/src/main/java/com/github/studydistractor/sdp/account/CreateAccountActivity.kt:122]: Please add a content description for accessibility

File: app/src/main/java/com/github/studydistractor/sdp/account/CreateAccountActivity.kt [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+                    Icon(Icons.Filled.Phone, contentDescription = null)
+                },
+                modifier = Modifier
+                    .fillMaxWidth()
+                    .padding(vertical = 8.dp)
+                    .testTag("phone")
+            )
+
+            // Validate
+            Button( onClick = {
```
Comment (aizensoosuke) [inline app/src/main/java/com/github/studydistractor/sdp/account/CreateAccountActivity.kt:181]: The way to do input validation in Jetpack Compose is like this : https://proandroiddev.com/input-validation-in-jetpack-compose-e99c18b44fe3

File: app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt [hunk 6]
```diff
@@ -0,0 +1,73 @@
+package com.github.studydistractor.sdp.account
+
+import android.util.Log
```
Comment (aizensoosuke) [inline app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt:3]: unused

File: app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+import android.widget.Toast
```
Comment (aizensoosuke) [inline app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt:4]: unused

File: app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+import androidx.compose.Context
```
Comment (aizensoosuke) [inline app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt:5]: unused

File: app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+import androidx.compose.NullCompilationScope
```
Comment (aizensoosuke) [inline app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt:6]: unused

File: app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+ * A class that implements the CreateAccInterface and provides methods to check if user input data
+ * is valid and to create an account in Firebase database.
+ */
+class FirebaseCreateAccount @Inject constructor(): CreateAccInterface{
+
+    private var auth = FirebaseAuth.getInstance()
+
+    private var db: FirebaseDatabase = FirebaseDatabase.getInstance()
+
+    private var PATH = "users"
```
Comment (aizensoosuke) [inline app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt:21]: unused

File: app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (aizensoosuke) [reply app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt:54]: We definitely need to restructure our code and be **consistent**. So, if we plan to do this for any other piece of data (which is a good practice), we should do it as soon as possible.

IMO you should only resolve if you don't have time to do it.

[the remark this answers is not in the document]
Comment (aizensoosuke) [reply app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt:54]: Actually it's just a useless risk. This method is not private, so please do error checking in it.

File: app/src/main/java/com/github/studydistractor/sdp/login/LoginActivity.kt [hunk 11]
```diff
@@ -113,7 +113,6 @@ class LoginActivity : Hilt_LoginActivity() {
                             } else {
                                 Toast.makeText(this@LoginActivity, "Login Failed", Toast.LENGTH_SHORT).show()
                                 Log.d(TAG, "signInWithEmail:failed")
-                                Log.d(TAG, "email: " + emailstr + ", password: " + passwordstr)
```
Comment (aizensoosuke) [inline app/src/main/java/com/github/studydistractor/sdp/login/LoginActivity.kt:116]: Actually this is useful, you should keep it

File: app/src/main/java/com/github/studydistractor/sdp/register/RegisterActivity.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (aizensoosuke) [reply app/src/main/java/com/github/studydistractor/sdp/register/RegisterActivity.kt:163]: There should be some idea of "steps" in your naming. With the current naming it sounds like you creating the account twice, and it's just not clear.

File: app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+    /**
+     * Check if the given phone number is not empty and its length is exactly 10 characters
+     * @param phoneNumber The phone number to check
+     * @return true if the phone number is correctly formatted
+     */
+    @Inject
+    override fun checkPhoneFormat(phoneNumber: String): Boolean {
```
Comment (aizensoosuke) [inline app/src/main/java/com/github/studydistractor/sdp/account/FirebaseCreateAccount.kt:29]: What about international numbers with +33 or +41

File: app/src/main/java/com/github/studydistractor/sdp/register/RegisterActivity.kt
Comment (aizensoosuke) [inline app/src/main/java/com/github/studydistractor/sdp/register/RegisterActivity.kt:1]: Please use material3 in here

**Review by aizensoosuke — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/github/studydistractor/sdp/login/LoginActivity.kt (hunk 11 above)
Comment (abouquet27) [reply app/src/main/java/com/github/studydistractor/sdp/login/LoginActivity.kt:116]: Not sure to let the password in the console is a good idea but I'll keep it then

File: app/src/androidTest/java/com/github/studydistractor/sdp/account/FakeCreateAccModule.kt [hunk 13]
```diff
@@ -0,0 +1,44 @@
+package com.github.studydistractor.sdp.account
+
+import androidx.compose.Context
```
[the remark this answers is not in the document]
Comment (abouquet27) [reply app/src/androidTest/java/com/github/studydistractor/sdp/account/FakeCreateAccModule.kt:3]: No it is used, check the wiki on Hilt

**Comment by aizensoosuke** [comment]
Of course, considering I did this review at the last minute, and considering it's a lot more detailed than what we were used to (but from now on I will do real reviews), I don't expect all of this to be resolved for this PR and I realize it might be a little overwhelming. 

One of the tasks for next print will be refactoring and, among other things, resolving all these comments (even if the PR is closed).

File: app/src/androidTest/java/com/github/studydistractor/sdp/account/FakeCreateAccModule.kt [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+    override fun postData(
+        year: Int,
+        month: Int,
+        day: Int,
+        firstname: String,
+        lastname: String,
+        phone: String
+    ) {
+        // Do Nothing
```
[the remark this answers is not in the document]
Comment (abouquet27) [reply app/src/androidTest/java/com/github/studydistractor/sdp/account/FakeCreateAccModule.kt:42]: Fake implementation such that the tests work (cf the wiki)

**Review by bluedrack — DISMISSED** [review]
LGTM. But pay attention at what Yohan as highlight.

File: app/src/main/java/com/github/studydistractor/sdp/login/LoginActivity.kt (hunk 11 above)
Comment (aizensoosuke) [reply app/src/main/java/com/github/studydistractor/sdp/login/LoginActivity.kt:116]: Actually it only goes to the console in debug mode

**Comment by abouquet27** [comment]
> LGTM. But pay attention at what Yohan as highlight.

Absolutely, it will be taken into account. All the refactoring will be denoted with issue #65

**Review by aizensoosuke — APPROVED** [review]
You can merge, but I expect all the comments in #65 to be resolved next sprint.
